controller/controller.go

```diff
@@ -294,11 +274,122 @@ func (c *Controller) Mount(mountRequest k8sresources.FlexVolumeMountRequest) k8s
+		Status:  FlexFailureStr,
+		Message: additionalMsg + err.Error(),
+	}
+	c.logger.Error(fmt.Sprintf("%#v", response))
+	return response
+}
+
+func (c *Controller) checkSlinkBeforeUmount(k8sPVDirectoryPath string, realMountedPath string) (bool, error) {
+	/*
+	   Return <isSlink, error>
+	   true,  nil   : slink exist as expected
+	   false, nil   : slink not exist, which indicate for idempotent issue
+	   true,  error : its slink but error during evaluate it or its slink but not to the right place
+	   false, error : its something different then slink
+	*/
+	defer c.logger.Trace(logs.DEBUG)()
+
+	// Identify the PV directory by using Lstat and then handle all idempotent cases (use Lstat to get the dir or slink detail and not the evaluation of it)
+	fileInfo, err := c.exec.Lstat(k8sPVDirectoryPath)
+	if err != nil {
+		if c.exec.IsNotExist(err) {
+			// The k8s PV directory not exist (its a rare case and indicate on idempotent flow)
+			c.logger.Info("PV directory(k8s-mountpoint) does not exist.", logs.Args{{"k8s-mountpoint", k8sPVDirectoryPath}, {"should-point-to-mountpoint", realMountedPath}}) // TODO change it to warning
+			return false, nil                                                                                                                                                 // Idempotent flow
+		} else {
+			// Maybe some permissions issue
+			return false, c.logger.ErrorRet(err, "Controller: failed to identify PV directory(k8s-mountpoint)", logs.Args{{"k8s-mountpoint", k8sPVDirectoryPath}})
+		}
+	} else if c.exec.IsSlink(fileInfo) {
+		// Its already slink so check if slink is ok and skip else raise error
+		evalSlink, err := c.exec.EvalSymlinks(k8sPVDirectoryPath)
+		if err != nil {
+			return true, c.logger.ErrorRet(err, "Controller: Idempotent - failed eval the slink of PV directory(k8s-mountpoint)", logs.Args{{"k8s-mountpoint", k8sPVDirectoryPath}})
+		}
+		if evalSlink == realMountedPath {
+			c.logger.Info("PV directory(k8s-mountpoint) is slink that point to the right mountpoint.", logs.Args{{"k8s-mountpoint", k8sPVDirectoryPath}, {"mountpoint", realMountedPath}})
+			return true, nil // Good path
+		} else {
+			// Very edge case, where the slink is point to wrong mountpoint
+			return true, c.logger.ErrorRet(
+				&wrongSlinkError{slink: k8sPVDirectoryPath, wrongPointTo: evalSlink, expectedPointTo: realMountedPath},
+				"failed")
+		}
+	} else {
+		return false, c.logger.ErrorRet(&k8sPVDirectoryIsNotSlinkError{k8sPVDirectoryPath, fileInfo}, "failed")
+	}
+
+}
+
+func (c *Controller) getRealMountpointForPvByBackend(volumeBackend string, volumeConfig map[string]interface{}) (string, error) {
+	// TODO we should create agnostic function based on backend that returns the real mountpoint
+	defer c.logger.Trace(logs.DEBUG)()
+	if volumeBackend == resources.SCBE {
+		return fmt.Sprintf(resources.PathToMountUbiquityBlockDevices, volumeConfig["Wwn"].(string)), nil
+	} else if volumeBackend == resources.SpectrumScale {
```

Comment:
I believe this would always return empty string for Spectrum Scale backend and unmount would fail. This has to be taken care.